File: templates/claude-code/.agentid/agentid-bootstrap.py
#!/usr/bin/env python3
"""
Agent ID cold-start bootstrap script.

Designed to be called:
  - automatically via a Claude Code SessionStart hook before the first turn
  - manually when you want to refresh state

Outputs the prompt block text to stdout so CLAUDE.md can instruct Claude to
read it, and writes all state artifacts to disk so they survive restarts.

Usage:
    python3 .agentid/agentid-bootstrap.py

Exit codes:
    0  - success (online or offline-fallback)
    1  - no identity found (no state.json, no env var)
"""

from __future__ import annotations

import json
import os
import sys
import hashlib
import datetime
import urllib.request
import urllib.error

# ── Config ────────────────────────────────────────────────────────────────────

AGENTID_DIR  = os.path.join(os.path.dirname(os.path.abspath(__file__)))
STATE_FILE   = os.path.join(AGENTID_DIR, "state.json")
API_BASE     = "https://getagent.id/api/v1"
USER_AGENT   = "AgentID-Client/1.0 claude-code/1.0"
TIMEOUT_SECS = 10

# ── Credential discovery ──────────────────────────────────────────────────────

def load_credentials() -> tuple[str, str]:
    """
    Returns (agent_id, api_key).

    Precedence:
      1. AGENTID_API_KEY env var  +  AGENTID_AGENT_ID env var
      2. ~/.agentid/state.json
      3. <workspace>/.agentid/state.json

    Raises SystemExit(1) if neither found.
    """
    env_key      = os.environ.get("AGENTID_API_KEY", "").strip()
    env_agent_id = os.environ.get("AGENTID_AGENT_ID", "").strip()

    # Try env first (best for CI / hosted agents)
    if env_key and env_agent_id:
        return env_agent_id, env_key

    # Try workspace state file
    for candidate in [STATE_FILE, os.path.expanduser("~/.agentid/state.json")]:
        if os.path.exists(candidate):
            try:
                with open(candidate, encoding="utf-8") as f:
                    state = json.load(f)
                agent_id = state.get("agent_id") or state.get("agentId", "")
                api_key  = (
                    env_key                       # honour env key if only agent_id is in state
                    or state.get("api_key", "")
                    or state.get("apiKey", "")
                )
                if agent_id and api_key:
                    return agent_id, api_key
            except Exception as exc:
                _warn(f"Could not read state file {candidate}: {exc}")

    print("[agentid-bootstrap] FATAL: No agent_id/api_key found.", file=sys.stderr)
    print(
        "  Set AGENTID_API_KEY + AGENTID_AGENT_ID env vars, "
        "or ensure .agentid/state.json exists.",
        file=sys.stderr,
    )
    sys.exit(1)


# ── HTTP helper ───────────────────────────────────────────────────────────────

def _api(method: str, path: str, api_key: str, body: dict | None = None) -> dict:
    """Make an authenticated request to the Agent ID API."""
    url = f"{API_BASE}{path}"
    data = json.dumps(body).encode() if body is not None else None
    req  = urllib.request.Request(
        url,
        data=data,
        method=method,
        headers={
            "Content-Type": "application/json",
            "Accept":       "application/json",
            "X-Agent-Key":  api_key,
            "User-Agent":   USER_AGENT,
        },
    )
    with urllib.request.urlopen(req, timeout=TIMEOUT_SECS) as resp:
        return json.loads(resp.read().decode())


def _warn(msg: str) -> None:
    print(f"[agentid-bootstrap] WARN  {msg}", file=sys.stderr)


def _info(msg: str) -> None:
    print(f"[agentid-bootstrap] INFO  {msg}", file=sys.stderr)


# ── Persistence ───────────────────────────────────────────────────────────────

def _write(filename: str, data: dict) -> None:
    os.makedirs(AGENTID_DIR, exist_ok=True)
    path = os.path.join(AGENTID_DIR, filename)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2)


def _read(filename: str) -> dict | None:
    path = os.path.join(AGENTID_DIR, filename)
    if not os.path.exists(path):
        return None
    try:
        with open(path, encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return None


# ── Bootstrap sequence ────────────────────────────────────────────────────────

def bootstrap() -> dict:
    """
    Full cold-start sequence. Returns a result dict.

    On any individual step failure the script continues with local cache
    so the agent still boots with the last known-good identity.
    """
    ts       = datetime.datetime.utcnow().isoformat() + "Z"
    stale    = False
    reasons: list[str] = []

    agent_id, api_key = load_credentials()
    _info(f"Agent: {agent_id}")

    result: dict = {
        "agent_id":       agent_id,
        "system_context": "",
        "prompt_block":   None,
        "heartbeat":      None,
        "marketplace":    None,
        "bootstrap":      None,
        "stale":          False,
        "stale_reasons":  [],
        "refreshed_at":   ts,
    }

    # 1. Heartbeat ─────────────────────────────────────────────────────────────
    hb: dict = {}
    try:
        hb = _api("POST", f"/agents/{agent_id}/heartbeat", api_key, {})
        result["heartbeat"] = hb
        _write("heartbeat.json", {**hb, "_refreshed_at": ts})
        _info("Heartbeat OK")
    except Exception as exc:
        stale = True
        reasons.append(f"heartbeat: {exc}")
        _warn(f"Heartbeat failed — using cached state. ({exc})")
        hb = _read("heartbeat.json") or {}

    # 2. Prompt block ──────────────────────────────────────────────────────────
    pb: dict = {}
    try:
        pb = _api("GET", f"/agents/{agent_id}/prompt-block?format=structured", api_key)
        result["prompt_block"]   = pb
        result["system_context"] = pb.get("promptText") or pb.get("text") or ""
        _write("prompt-block.json", {**pb, "_refreshed_at": ts})
        _info(f"Prompt block fetched (checksum={pb.get('checksum', 'n/a')})")
    except Exception as exc:
        stale = True
        reasons.append(f"prompt-block: {exc}")
        _warn(f"Prompt block fetch failed — falling back to cache. ({exc})")
        pb = _read("prompt-block.json") or {}
        result["prompt_block"]   = pb
        result["system_context"] = pb.get("promptText") or pb.get("text") or ""

    # 3. Marketplace context ───────────────────────────────────────────────────
    alerts = hb.get("stateDelta", {}).get("marketplace_alerts", {})
    if alerts.get("any_action_required") or alerts.get("orders_requiring_acceptance", 0) > 0:
        try:
            mkt = _api("GET", f"/agents/{agent_id}/marketplace/context", api_key)
            result["marketplace"] = mkt
            _write("marketplace-context.json", {**mkt, "_refreshed_at": ts})
            _info("Marketplace context fetched (action required)")
        except Exception as exc:
            reasons.append(f"marketplace/context: {exc}")
            _warn(f"Marketplace context fetch failed. ({exc})")

    # 4. Bootstrap refresh (when signalled) ────────────────────────────────────
    if hb.get("stateDelta", {}).get("action_required") or not _read("bootstrap.json"):
        try:
            bs = _api("GET", f"/agents/{agent_id}/bootstrap", api_key)
            result["bootstrap"] = bs
            _write("bootstrap.json", {**bs, "_refreshed_at": ts})
            _info("Bootstrap bundle refreshed")
        except Exception as exc:
            reasons.append(f"bootstrap: {exc}")
            _warn(f"Bootstrap refresh failed. ({exc})")
            result["bootstrap"] = _read("bootstrap.json")

    # 5. Persist updated state.json ────────────────────────────────────────────
    state = _read("state.json") or {"version": 1}
    state.update({
        "agent_id":              agent_id,
        "did":                   f"did:web:getagent.id:agents:{agent_id}",
        "inbox":                 f"{agent_id}@getagent.id",
        "last_cold_start":       ts,
        "last_heartbeat_at":     hb.get("lastHeartbeatAt", ts),
        "prompt_block_checksum": pb.get("checksum"),
        "prompt_block_version":  pb.get("version"),
        "stale":                 stale,
        "stale_reasons":         reasons,
    })
    _write("state.json", state)

    result["stale"]        = stale
    result["stale_reasons"] = reasons
    return result


# ── Entry point ───────────────────────────────────────────────────────────────

def main() -> None:
    result = bootstrap()

    system_context = result.get("system_context", "")
    agent_id       = result.get("agent_id", "")

    if result["stale"]:
        print(
            f"[agentid-bootstrap] Running on stale/cached state: {result['stale_reasons']}",
            file=sys.stderr,
        )

    # Emit a brief identity summary to stderr (shows in Claude Code SessionStart output)
    print(
        f"[agentid-bootstrap] Identity loaded — "
        f"agent_id={agent_id} "
        f"stale={result['stale']}",
        file=sys.stderr,
    )

    # Print the prompt block text to stdout — Claude Code session hook sees this
    if system_context:
        print(system_context)
    else:
        print(
            f"[Agent ID identity — loaded from cache]\n"
            f"agentId: {agent_id}\n"
            f"did: did:web:getagent.id:agents:{agent_id}\n"
            f"inbox: {agent_id}@getagent.id\n"
            f"Note: Prompt block could not be fetched from network. Using last known state."
        )


if __name__ == "__main__":
    main()

File: templates/nodejs/agentid-bootstrap.mjs
#!/usr/bin/env node
/**
 * Agent ID cold-start bootstrap — pure Node.js stdlib, no dependencies.
 *
 * Works with any Node.js agent framework (LangChain.js, OpenAI SDK,
 * AutoGen.js, custom agents, etc.).
 *
 * Usage (add to your agent's startup):
 *   import { bootstrap, loadCredentials } from './agentid-bootstrap.mjs';
 *   const { systemContext } = await bootstrap();
 *   // → inject systemContext into your system prompt
 *
 * Or run standalone to refresh state files:
 *   node .agentid/agentid-bootstrap.mjs
 *
 * Credential discovery (in priority order):
 *   1. AGENTID_API_KEY + AGENTID_AGENT_ID env vars
 *   2. .agentid/state.json
 *   3. ~/.agentid/state.json
 */

import { createHash, randomBytes } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { homedir } from "node:os";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));

const AGENTID_DIR = resolve(__dirname);
const STATE_FILE = join(AGENTID_DIR, "state.json");
const API_BASE = "https://getagent.id/api/v1";
const USER_AGENT = "AgentID-Client/1.0 nodejs/agentid-bootstrap";
const TIMEOUT_MS = 10_000;

// ── Credential discovery ────────────────────────────────────────────────────

export function loadCredentials() {
  const envKey = (process.env.AGENTID_API_KEY || "").trim();
  const envAgentId = (process.env.AGENTID_AGENT_ID || "").trim();

  if (envKey && envAgentId) return { agentId: envAgentId, apiKey: envKey };

  const candidates = [STATE_FILE, join(homedir(), ".agentid", "state.json")];
  for (const candidate of candidates) {
    if (!existsSync(candidate)) continue;
    try {
      const state = JSON.parse(readFileSync(candidate, "utf-8"));
      const agentId = state.agent_id || state.agentId || "";
      const apiKey = envKey || state.api_key || state.apiKey || "";
      if (agentId && apiKey) return { agentId, apiKey };
    } catch {
      // continue to next candidate
    }
  }

  throw new Error(
    "[agentid-bootstrap] FATAL: No agent_id/api_key found.\n" +
    "  Set AGENTID_API_KEY + AGENTID_AGENT_ID env vars,\n" +
    "  or ensure .agentid/state.json exists."
  );
}

// ── HTTP helper ─────────────────────────────────────────────────────────────

async function apiRequest(method, path, apiKey, body = null) {
  const url = `${API_BASE}${path}`;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);

  try {
    const res = await fetch(url, {
      method,
      signal: controller.signal,
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
        "X-Agent-Key": apiKey,
        "User-Agent": USER_AGENT,
      },
      body: body !== null ? JSON.stringify(body) : undefined,
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${await res.text()}`);
    return await res.json();
  } finally {
    clearTimeout(timer);
  }
}

// ── Persistence ─────────────────────────────────────────────────────────────

function writeState(filename, data) {
  mkdirSync(AGENTID_DIR, { recursive: true });
  const path = join(AGENTID_DIR, filename);
  writeFileSync(path, JSON.stringify(data, null, 2), "utf-8");
}

function readState(filename) {
  const path = join(AGENTID_DIR, filename);
  if (!existsSync(path)) return null;
  try {
    return JSON.parse(readFileSync(path, "utf-8"));
  } catch {
    return null;
  }
}

// ── Bootstrap sequence ──────────────────────────────────────────────────────

/**
 * Execute the full cold-start sequence.
 *
 * @returns {{
 *   agentId: string,
 *   systemContext: string,
 *   promptBlock: object|null,
 *   heartbeat: object|null,
 *   marketplace: object|null,
 *   bootstrap: object|null,
 *   stale: boolean,
 *   staleReasons: string[],
 * }}
 */
export async function bootstrap() {
  const ts = new Date().toISOString();
  const staleReasons = [];

  const { agentId, apiKey } = loadCredentials();
  console.error(`[agentid-bootstrap] INFO  Agent: ${agentId}`);

  const result = {
    agentId,
    systemContext: "",
    promptBlock: null,
    heartbeat: null,
    marketplace: null,
    bootstrap: null,
    stale: false,
    staleReasons: [],
    refreshedAt: ts,
  };

  // 1. Heartbeat ─────────────────────────────────────────────────────────────
  let hb = {};
  try {
    hb = await apiRequest("POST", `/agents/${agentId}/heartbeat`, apiKey, {});
    result.heartbeat = hb;
    writeState("heartbeat.json", { ...hb, _refreshed_at: ts });
    console.error("[agentid-bootstrap] INFO  Heartbeat OK");
  } catch (err) {
    staleReasons.push(`heartbeat: ${err.message}`);
    console.error(`[agentid-bootstrap] WARN  Heartbeat failed — using cache. (${err.message})`);
    hb = readState("heartbeat.json") || {};
  }

  // 2. Prompt block ──────────────────────────────────────────────────────────
  let pb = {};
  try {
    pb = await apiRequest("GET", `/agents/${agentId}/prompt-block?format=structured`, apiKey);
    result.promptBlock = pb;
    result.systemContext = pb.promptText || pb.text || "";
    writeState("prompt-block.json", { ...pb, _refreshed_at: ts });
    console.error(`[agentid-bootstrap] INFO  Prompt block fetched (checksum=${pb.checksum || "n/a"})`);
  } catch (err) {
    staleReasons.push(`prompt-block: ${err.message}`);
    console.error(`[agentid-bootstrap] WARN  Prompt block fetch failed — falling back to cache. (${err.message})`);
    pb = readState("prompt-block.json") || {};
    result.promptBlock = pb;
    result.systemContext = pb.promptText || pb.text || "";
  }

  // 3. Marketplace context (only when action required) ───────────────────────
  const alerts = hb?.stateDelta?.marketplace_alerts || {};
  if (alerts.any_action_required || (alerts.orders_requiring_acceptance || 0) > 0) {
    try {
      const mkt = await apiRequest("GET", `/agents/${agentId}/marketplace/context`, apiKey);
      result.marketplace = mkt;
      writeState("marketplace-context.json", { ...mkt, _refreshed_at: ts });
      console.error("[agentid-bootstrap] INFO  Marketplace context fetched (action required)");
    } catch (err) {
      staleReasons.push(`marketplace/context: ${err.message}`);
      console.error(`[agentid-bootstrap] WARN  Marketplace context fetch failed. (${err.message})`);
    }
  }

  // 4. Bootstrap bundle (when signalled or missing) ──────────────────────────
  const needsBootstrap = hb?.stateDelta?.action_required || !readState("bootstrap.json");
  if (needsBootstrap) {
    try {
      const bs = await apiRequest("GET", `/agents/${agentId}/bootstrap`, apiKey);
      result.bootstrap = bs;
      writeState("bootstrap.json", { ...bs, _refreshed_at: ts });
      console.error("[agentid-bootstrap] INFO  Bootstrap bundle refreshed");
    } catch (err) {
      staleReasons.push(`bootstrap: ${err.message}`);
      console.error(`[agentid-bootstrap] WARN  Bootstrap refresh failed. (${err.message})`);
      result.bootstrap = readState("bootstrap.json");
    }
  } else {
    result.bootstrap = readState("bootstrap.json");
  }

  // 5. Persist state.json ────────────────────────────────────────────────────
  const state = readState("state.json") || { version: 1 };
  Object.assign(state, {
    agent_id: agentId,
    did: `did:web:getagent.id:agents:${agentId}`,
    inbox: `${agentId}@getagent.id`,
    last_cold_start: ts,
    last_heartbeat_at: hb.lastHeartbeatAt || ts,
    prompt_block_checksum: pb.checksum || null,
    prompt_block_version: pb.version || null,
    stale: staleReasons.length > 0,
    stale_reasons: staleReasons,
  });
  writeState("state.json", state);

  result.stale = staleReasons.length > 0;
  result.staleReasons = staleReasons;
  return result;
}

// ── Standalone entry point ──────────────────────────────────────────────────

if (process.argv[1] && resolve(process.argv[1]) === resolve(fileURLToPath(import.meta.url))) {
  const result = await bootstrap();

  if (result.stale) {
    console.error(`[agentid-bootstrap] Running on stale/cached state: ${result.staleReasons}`);
  }

  console.error(
    `[agentid-bootstrap] Identity loaded — ` +
    `agent_id=${result.agentId} stale=${result.stale}`
  );

  // Print prompt block to stdout (captured by process supervisors / hooks)
  if (result.systemContext) {
    process.stdout.write(result.systemContext + "\n");
  } else {
    process.stdout.write(
      `[Agent ID identity — loaded from cache]\n` +
      `agentId: ${result.agentId}\n` +
      `did: did:web:getagent.id:agents:${result.agentId}\n` +
      `inbox: ${result.agentId}@getagent.id\n` +
      `Note: Prompt block could not be fetched from network. Using last known state.\n`
    );
  }
}
